components/service/sync-logins/src/test/java/mozilla/components/service/sync/logins/DefaultLoginValidationDelegateTest.kt

```diff
@@ -0,0 +1,202 @@
+import org.robolectric.annotation.Config
+
+@RunWith(AndroidJUnit4::class)
+@ExperimentalCoroutinesApi
+class DefaultLoginValidationDelegateTest {
+
+    private lateinit var loginsStorage: LoginsStorage
+    private lateinit var validationDelegate: DefaultLoginValidationDelegate
+    private lateinit var scope: TestCoroutineScope
+
+    @Before
+    @Config(sdk = [21])
+    fun before() {
+        loginsStorage = mockLoginsStorage()
+        scope = TestCoroutineScope()
+        validationDelegate = DefaultLoginValidationDelegate(storage = loginsStorage, scope = scope)
+    }
+
+    @Test
+    @Config(sdk = [21])
+    fun `WHEN guid is null or empty THEN should create a new record`() = runBlockingTest {
+        val fromNull = validationDelegate.getPersistenceOperation(createLogin(guid = null)).await()
+        val fromEmpty = validationDelegate.getPersistenceOperation(createLogin(guid = "")).await()
+
+        assertEquals(Operation.CREATE, fromNull)
+        assertEquals(Operation.CREATE, fromEmpty)
+    }
+
+    @Test
+    @Config(sdk = [21])
+    fun `WHEN existing login matches by guid, should update`() = runBlockingTest {
+        val newLogin = createLogin(guid = "1")
+        whenever(loginsStorage.get(anyString())).thenReturn(
+            createLogin(
+                guid = "1",
+                password = "password1"
+            )
+        )
+        val operation = validationDelegate.getPersistenceOperation(newLogin).await()
+        assertEquals(Operation.UPDATE, operation)
+    }
+
+    @Test
+    @Config(sdk = [21])
+    fun `WHEN saved login GUID matches, and is missing a username, should update`() =
+        runBlockingTest {
+            val newLogin = createLogin(guid = "1")
+            whenever(loginsStorage.get(anyString())).thenReturn(
+                createLogin(
+                    guid = "1",
+                    username = ""
+                )
+            )
+            val operation = validationDelegate.getPersistenceOperation(newLogin).await()
+            assertEquals(Operation.UPDATE, operation)
+        }
+
+    @Test
+    @Config(sdk = [21])
+    fun `WHEN saved login GUID matches, but username is updated, we should create`() =
+        runBlockingTest {
+            val newLogin = createLogin(guid = "1")
+            whenever(loginsStorage.get(anyString())).thenReturn(
+                createLogin(
+                    guid = "1",
+                    username = "username2"
+                )
+            )
+            val operation = validationDelegate.getPersistenceOperation(newLogin).await()
+            assertEquals(Operation.CREATE, operation)
+        }
+
+    @Test
+    @Config(sdk = [21])
+    fun `WHEN existing login matches by important elements, should update`() = runBlockingTest {
+        val newLogin = createLogin(guid = "1")
+        val savedLogin = createLogin("2")
+        whenever(loginsStorage.getStoredLoginByElements(any())).thenReturn(savedLogin)
```

Comment:
This fails locally (and probably on CI too) with an NPE. :/ Not sure what I'm doing wrong. Clearly since the other new tests work it's something with the `any()`

Reply:
> java.lang.NullPointerException at mozilla.components.service.sync.logins.DefaultLoginValidationDelegate$Companion.getStoredLoginByElements(DefaultLoginValidationDelegate.kt:105)
> 	at mozilla.components.service.sync.logins.DefaultLoginValidationDelegateTest$WHEN existing login matches by important elements, should update$1.invokeSuspend(DefaultLoginValidationDelegateTest.kt:100)
> 	at mozilla.components.service.sync.logins.DefaultLoginValidationDelegateTest$WHEN existing login matches by important elements, should update$1.invoke(DefaultLoginValidationDelegateTest.kt)
> 	at kotlinx.coroutines.test.TestBuildersKt$runBlockingTest$deferred$1.invokeSuspend(TestBuilders.kt:50)
> 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
> 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
> 	at kotlinx.coroutines.test.TestCoroutineDispatcher.dispatch(TestCoroutineDispatcher.kt:50)
> 	at kotlinx.coroutines.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:271)
> 	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:26)
> 	at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:109)
> 	at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:158)
> 	at kotlinx.coroutines.BuildersKt__Builders_commonKt.async(Builders.common.kt:89)
> 	at kotlinx.coroutines.BuildersKt.async(Unknown Source)
> 	at kotlinx.coroutines.BuildersKt__Builders_commonKt.async$default(Builders.common.kt:82)
> 	at kotlinx.coroutines.BuildersKt.async$default(Unknown Source)
> 	at kotlinx.coroutines.test.TestBuildersKt.runBlockingTest(TestBuilders.kt:49)
> 	at kotlinx.coroutines.test.TestBuildersKt.runBlockingTest$default(TestBuilders.kt:45)
> 	at mozilla.components.service.sync.logins.DefaultLoginValidationDelegateTest.WHEN existing login matches by important elements, should update(DefaultLoginValidationDelegateTest.kt:97)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
> 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
> 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
> 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
> 	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:601)
> 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
> 	at org.robolectric.internal.SandboxTestRunner$2.evaluate(SandboxTestRunner.java:260)
> 	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:130)
> 	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:42)
> 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
> 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
> 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
> 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
> 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
> 	at org.robolectric.internal.SandboxTestRunner$1.evaluate(SandboxTestRunner.java:84)
> 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
> 	at androidx.test.ext.junit.runners.AndroidJUnit4.run(AndroidJUnit4.java:104)
> 	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
> 	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
> 	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
> 	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
> 	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)